src/data/sentencepiece_vocab.h

```diff
@@ -0,0 +1,96 @@
+
+#ifdef USE_SENTENCEPIECE
+#include "sentencepiece/src/sentencepiece_processor.h"
+#include "sentencepiece/src/sentencepiece_trainer.h"
+#endif
+
+#include "common/config.h"
+#include "common/filesystem.h"
+#include "common/logging.h"
+#include "common/options.h"
+#include "common/regex.h"
+
+#include <random>
+#include <sstream>
+
+
+
+namespace sentencepiece {
+  class SentencePieceText;
+}
+
+namespace marian {
+#ifdef USE_SENTENCEPIECE
+
+// Wrapper around https://github.com/google/sentencepiece
+class SentencePieceVocab : public IVocab {
+private:
+  // Actual SentencePiece processor object
+  UPtr<sentencepiece::SentencePieceProcessor> spm_;
+
+  // Sampling factor for subword regularization, disabled when 0
+  float alpha_{0};
+
+  // Allowed suffixes for SentencePiece model
+  std::vector<std::string> suffixes_ = {".spm"};
+
+  Ptr<Options> options_;
+  size_t batchIndex_{0};
+
+  std::mt19937 generator_;
+  std::uniform_int_distribution<int> randInt_;  // from 0 to INT_MAX
+
+  // Keeps sentences segmented into subword units
+  bool keepEncoded_{false};
+
+  // Sample from one file, based on first algorithm from:
+  // https://en.wikipedia.org/wiki/Reservoir_sampling
+  void reservoirSampling(std::vector<std::string>& sample,
+                         size_t& seenLines,
+                         const std::string& trainPath,
+                         size_t maxLines,
+                         size_t maxBytes);
+
+  // Iterate over all input files and collect a representative sample via reservoir sampling.
+  // The sample will first grow to the desired size and next keep sampling with decreasing
+  // probability in the hope to get a uniform sample from the union of all files.
+  size_t reservoirSamplingAll(io::TemporaryFile& temp,
+                              const std::vector<std::string>& trainPaths,
+                              size_t maxLines,
+                              size_t maxBytes);
+
+  // Just concatenate all files to a temporary file so SentencePiece can consume it.
+  size_t dumpAll(io::TemporaryFile& temp,
+                 const std::vector<std::string>& trainPaths,
+                 size_t maxBytes);
+
+public:
+  SentencePieceVocab(Ptr<Options> options, size_t batchIndex);
+
+  virtual const std::string& canonicalExtension() const override;
+  virtual const std::vector<std::string>& suffixes() const override;
+
+  virtual std::string suffix();
+  virtual std::string type() const override;
+  virtual Word getEosId() const override;
+  virtual Word getUnkId() const override;
+
+  void create(const std::string& vocabPath,
+              const std::vector<std::string>& trainPaths,
+              size_t maxSize) override;
+  void createFake() override;
+  Word operator[](const std::string& token) const override;
+  const std::string& operator[](Word id) const override;
+  Words encode(const std::string& line, bool addEOS, bool inference) const override;
+  void encodeAsSentencePieceText(const std::string& line,
```

Comment:
Why not just make a default implementation that throws/aborts/errors in the virtual parent?  Then you wouldn't need to move everything to the header.  And you're already doing a dynamic cast to get here, which is arguably worse.  
Even better would be processing into whatever byte aligned data structure you need, to include processing SentencePieceText within the method here and never exposing it.

Reply:
Honestly, I have been trying to avoid the dynamic cast (which is already in the code, probably for `FactoredVocab`), which is what nearly cost me a day. But this method is only SentencePieceVocab as it stands now. 

For what you're proposing, is the intention to future-proof normalizing nature of other possible SentencePiece alternatives?
In that case, I can abstract the parent to return a common implementation for the byte-aligned data structure (for marian vocabs) restricting sentencepiece dependency to just the SenteneceVocab, which I think is an additional advantage? 

In terms of naming `encodeAsSentencePieceText` -> `encodeWithStringRef`?